codalab/rest/bundles.py

```diff
@@ -451,16 +458,71 @@ def _fetch_bundle_locations(bundle_uuid: str):
     """
-    Adds a new BundleLocation to a bundle. Request body must contain the fields in BundleLocationSchema.
+    Adds a new BundleLocation to a bundle. If need to generate sas token, generate Azure SAS token and connection string. Request body must contain the fields in BundleLocationSchema.
 
     Query parameters:
-    - `bundle_uuid`: Bundle UUID corresponding to the new location
+    - `need_sas`: (Optional) Bool. If true, if will return SAS token
+    - `is_dir`: (Optional) Bool. Whether the uploaded file is directory.
     """
+    check_bundles_have_all_permission(local.model, request.user, [bundle_uuid])
+    need_sas = query_get_bool('need_sas', default=False)
+    is_dir = query_get_bool('is_dir', default=False)
+
+    bundle = local.model.get_bundle(bundle_uuid)
     new_location = BundleLocationSchema(many=True).load(request.json).data[0]
-    new_location["uuid"] = local.model.add_bundle_location(
-        new_location['bundle_uuid'], new_location['bundle_store_uuid']
+    logging.info(
+        f"Try to bypass server upload, need_sas: {need_sas}, is_dir: {is_dir}, new_location: {new_location}"
     )
-    return BundleLocationSchema(many=True).dump([new_location]).data
+    if new_location.get('bundle_store_uuid', None) is None:
+        if os.environ.get('CODALAB_ALWAYS_USE_AZURE_BLOB_BETA') == '1':
+            # The rest-server use Azure as default storage. Use default azure path
+            local.model.update_bundle(
+                bundle, {'storage_type': StorageType.AZURE_BLOB_STORAGE.value, 'is_dir': is_dir},
+            )
+            bundle_url = local.bundle_store.get_bundle_location(bundle_uuid)
+        else:  # The rest-server does not support store on Azure.
+            bundle_url = None
+    else:
+        local.model.add_bundle_location(
+            new_location['bundle_uuid'], new_location['bundle_store_uuid']
+        )
+        local.model.update_bundle(  # get_bundle_location() function uses this field
+            bundle, {'is_dir': is_dir},
+        )
+        # TODO: check whether this function could get bundle location for GCS store.
+        bundle_url = local.bundle_store.get_bundle_location(bundle_uuid)
+        logging.info(f"Bypass server upload, url: {bundle_url}")
+    data = BundleLocationSchema(many=True).dump([new_location]).data
+
+    if need_sas:
+        if bundle_url is None:
+            bundle_conn_str, index_conn_str = None, None
+        elif bundle_url.startswith(StorageURLScheme.AZURE_BLOB_STORAGE.value):
+            # generate the SAS token and Azure connection string, and send it back to the client
+            bundle_sas_token = local.upload_manager.get_bundle_sas_token(bundle_url)
+            index_sas_token = local.upload_manager.get_index_sas_token(bundle_url)
+            base_conn_str = get_base_conn_str()
+            if LOCAL_USING_AZURITE and get_request_source() == RequestSource.CLI:
+                # For test locally. Mannually typing `cl upload`
```

Comment:
```suggestion
                # For test locally. Manually typing `cl upload`
```

I don't understand this, we should discuss